### PR TITLE
Initial commit of inference SDK for Bunkerhill Health

### PR DESCRIPTION
**Pull request by jackpaparian**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+`ModelRunner`. An example of a simple model class is shown below:
+
+```python
+from typing import Dict
+
+import numpy as np
+
+from tensorflow import keras
+
+from bunkerhill.bunkerhill_types import OutputAttributes, SeriesUID
```
Comment (gnedivad) [inline README.md]: Rename type from `SeriesUID` to `SeriesInstanceUID`.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+[`ModelRunner`](bunkerhill/model_runner.py) hosts each model in a gRPC server and runs inference
+in response to client requests. The sequence of interactions between client and server are as
+follows:
+
+1. Client saves pickled inference arguments to a file
+2. Client sends an [`InferenceRequest`](bunkerhill/proto/inference.proto#L14) to `ModelRunner`
+3. Upon receipt of the `InferenceRequest`, `ModelRunner` loads the arguments, runs inference, and
+saves the pickled output to a file
+4. `ModelRunner` sends an [`InferenceResponse`](bunkerhill/proto/inference.proto#L23) back to the
+client
```
Comment (gnedivad) [inline README.md:128]: These details should eventually be abstracted away from our users.

File: README.md [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+
+## Output attributes
+
+`inference()` can return any type(s) of output attributes. Examples include:
+
+- Segmentation arrays
+- Softmax arrays
+- Predicted classifications
+- Predicted scores
+
```
Comment (gnedivad) [inline README.md]: Remove extra newline.

File: README.md [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+from bunkerhill.model import BaseModel
+from bunkerhill.model_runner import ModelRunner
+
+class MyModel(BaseModel):
+
+  def __init__(self):
+    # This path must be valid within the Docker container.
+    self.model = keras.models.load_model('/path/to/weights')
+
+  def inference(self, pixel_array: Dict[SeriesUID, np.ndarray]) -> OutputAttributes:
```
Comment (gnedivad) [inline README.md]: Update this stub to include all of the other available study input attributes so that it's clearer what types each of them have.

File: bunkerhill/base_model.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import abc
+
+from bunkerhill.bunkerhill_types import OutputAttributes
+
+
+class BaseModel(abc.ABC):
+  """The base interface for a model that support inference()"""
+
+  @abc.abstractmethod
+  def inference(self, **kwargs) -> OutputAttributes:
```
Comment (gnedivad) [inline bunkerhill/base_model.py]: Nit: type hint `kwargs: Any`.

File: README.md [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+- [`RescaleIntercept`](https://dicom.innolitics.com/ciods/digital-x-ray-image/dx-image/00281052)
+- [`RescaleSlope`](https://dicom.innolitics.com/ciods/digital-x-ray-image/dx-image/00281053)
+- [`SliceThickness`](https://dicom.innolitics.com/ciods/rt-dose/image-plane/00180050)
+- [`WindowCenter`](https://dicom.innolitics.com/ciods/digital-x-ray-image/dx-image/00281050)
+- [`WindowWidth`](https://dicom.innolitics.com/ciods/digital-x-ray-image/dx-image/00281051)
+
+Other input attributes can be added to support new models. Please contact us if your model requires
+other input attributes for inference.
+
+## Output attributes
```
Comment (gnedivad) [inline README.md]: Should we call these outputs to be consistent with the language of inputs?

File: README.md [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+
+`inference()` can return any type(s) of output attributes. Examples include:
```
Comment (gnedivad) [inline README.md]: `inference()` returns a `Dict[str, Any]`, where each key-value pair represents a single output. The key is a unique name for the output and the value is the corresponding value of the output. Examples of values include:

File: bunkerhill/base_model.py (hunk 5 above)
Comment (gnedivad) [reply bunkerhill/base_model.py]: See: comment about renaming output attributes to outputs.

File: README.md [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+
+Inputs to your model's `inference()` method can either be specific to the DICOM study or a specific
+series in the study. Inputs specific to a series are passed to `inference()` as a dictionary
+mapping the DICOM series UUID to the input value, while inputs specific to a study are
+just passed to `inference()` by their value.
+
+### `PixelArray`
+
+DICOM files store their pixel data in the
+[`PixelData`](https://dicom.innolitics.com/ciods/segmentation/image-pixel/7fe00010) attribute.
```
Comment (gnedivad) [inline README.md]: attribute => data element

File: README.md [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+[pydicom `PixelArray` API](https://pydicom.github.io/pydicom/dev/old/working_with_pixel_data.html).
+
+For a given DICOM series, Bunkerhill concatentates each slice's array into a
+single 3D array.
+
+### Non-`PixelArray` inputs
+
+In addition to `PixelArray`, other DICOM attributes can be used as model inputs. These input
+attributes can be either be defined in the DICOM spec or can be defined as
+[private data elements](https://pydicom.github.io/pydicom/dev/old/private_data_elements.html).
```
Comment (gnedivad) [inline README.md]: "In addition to `PixelArray` other DICOM standard data elements can be used as model inputs."

Don't think we should support private data elements in this first version.

File: README.md [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+- [`PatientSex`](https://dicom.innolitics.com/ciods/arterial-pulse-waveform/patient/00100040)
+- [`PhotometricInterpretation`](https://dicom.innolitics.com/ciods/cr-image/cr-image/00280004)
+- [`RescaleIntercept`](https://dicom.innolitics.com/ciods/digital-x-ray-image/dx-image/00281052)
+- [`RescaleSlope`](https://dicom.innolitics.com/ciods/digital-x-ray-image/dx-image/00281053)
+- [`SliceThickness`](https://dicom.innolitics.com/ciods/rt-dose/image-plane/00180050)
+- [`WindowCenter`](https://dicom.innolitics.com/ciods/digital-x-ray-image/dx-image/00281050)
+- [`WindowWidth`](https://dicom.innolitics.com/ciods/digital-x-ray-image/dx-image/00281051)
+
+Other input attributes can be added to support new models. Please contact us if your model requires
+other input attributes for inference.
```
Comment (gnedivad) [inline README.md]: Other inputs can be added to support new models. Please contact us if your model requires other inputs for inference.

File: README.md [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+  def __init__(self):
+    # This path must be valid within the Docker container.
+    self.model = keras.models.load_model('/path/to/weights')
+
+  def inference(self, pixel_array: Dict[SeriesUID, np.ndarray]) -> OutputAttributes:
+    """Runs inference on an entire DICOM series.
+
+    Args:
+      pixel_array: A Dict mapping the DICOM series UID to the 3D tensor of concatenated PixelData
+        data elements.
```
Comment (gnedivad) [inline README.md]: ...the DICOM series instance UID to...

File: bunkerhill/bunkerhill_types.py [hunk 12]
```diff
@@ -0,0 +1,7 @@
+"""Types exposed to bunkerhill package."""
+
+from typing import Any, Dict
+
+OutputAttributes = Dict[str, Any]
+SeriesUID = str
```
Comment (gnedivad) [inline bunkerhill/bunkerhill_types.py]: See: comment about renaming type from `SeriesUID` to `SeriesInstanceUID`.

File: bunkerhill/bunkerhill_types.py [hunk 13]
```diff
@@ -0,0 +1,7 @@
+"""Types exposed to bunkerhill package."""
+
+from typing import Any, Dict
+
+OutputAttributes = Dict[str, Any]
```
Comment (gnedivad) [inline bunkerhill/bunkerhill_types.py]: See: comment about renaming `OutputAttributes` to `Outputs`.

File: bunkerhill/examples/hippocampus/Dockerfile [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+  && ln -s cuda-11.7 /usr/local/cuda \
+  && rm -rf /var/lib/apt/lists/*
+RUN echo "/usr/local/nvidia/lib" >> /etc/ld.so.conf.d/nvidia.conf \
+  && echo "/usr/local/nvidia/lib64" >> /etc/ld.so.conf.d/nvidia.conf
+
+# Installs Python 3.9
+RUN apt update -y \
+  && apt install -y software-properties-common
+RUN DEBIAN_FRONTEND="noninteractive" add-apt-repository ppa:deadsnakes/ppa
+RUN DEBIAN_FRONTEND="noninteractive" apt install -y python3.9
```
Comment (gnedivad) [inline bunkerhill/examples/hippocampus/Dockerfile]: `python3.9-dev`

File: bunkerhill/examples/hippocampus/Dockerfile [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+COPY README.md README.md
+COPY setup.py setup.py
+RUN pip install --editable .
+
+ENV PYTHONUNBUFFERED=1
+
+# Add a new user "host_user"
+RUN useradd -u ${USER_ID} host_user
+# Change to non-root privilege
+USER host_user
```
Comment (gnedivad) [inline bunkerhill/examples/hippocampus/Dockerfile:99]: Call out what the user needs to configure here.

File: bunkerhill/examples/hippocampus/conftest.py [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+@pytest.fixture(scope="function")
+def model() -> ModelRunner:
+  return ModelRunner(model_release=MSDHippocampusModel())
+
+
+@pytest.fixture(scope="function")
+def grpc_server(model) -> Server:
+  service = DESCRIPTOR.services_by_name['InferenceProcessor']
+  return grpc_testing.server_from_dictionary({service: model}, grpc_testing.strict_real_time())
+
```
Comment (gnedivad) [inline bunkerhill/examples/hippocampus/conftest.py]: Remove extra newline.

File: bunkerhill/examples/hippocampus/Dockerfile [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+
```
Comment (gnedivad) [inline bunkerhill/examples/hippocampus/Dockerfile]: Remove extra newline.

File: bunkerhill/examples/hippocampus/requirements.txt [hunk 18]
```diff
@@ -0,0 +1,5 @@
+matplotlib==3.6.3
+nnunet==1.7.0
+pytest==7.2.1
+torch==1.13.1
+
```
Comment (gnedivad) [inline bunkerhill/examples/hippocampus/requirements.txt]: Remove extra newline.

File: bunkerhill/examples/hippocampus/requirements.txt [hunk 19]
```diff
@@ -0,0 +1,5 @@
+matplotlib==3.6.3
```
Comment (gnedivad) [inline bunkerhill/examples/hippocampus/requirements.txt:1]: Why is `matplotlib` a requirement in this example?

File: bunkerhill/examples/hippocampus/test_model.py [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+    outputs = pickle.load(f)
+
+  segmentation = outputs[MSDHippocampusModel._SEGMENTATION_OUTPUT_ATTRIBUTE_NAME][series_uid]
+  assert segmentation.shape == (39, 47, 36)
+  assert segmentation.dtype == np.uint8
+
+  softmax = outputs[MSDHippocampusModel._SOFTMAX_OUTPUT_ATTRIBUTE_NAME][series_uid]
+  assert softmax.shape == (3, 36, 47, 39)
+  assert softmax.dtype == np.float16
+
```
Comment (gnedivad) [inline bunkerhill/examples/hippocampus/test_model.py]: Remove extra newline.

File: bunkerhill/grpc_socket.py [hunk 21]
```diff
@@ -0,0 +1,4 @@
+"""Unix domain socket for gRPC service."""
+
+UNIX_DOMAIN_SOCKET_PATH_TEMPLATE = 'unix://%s/model_runner.sock'
+
```
Comment (gnedivad) [inline bunkerhill/grpc_socket.py]: Remove extra newline.

**Review by gnedivad — APPROVED** [review]

File: bunkerhill/examples/hippocampus/requirements.txt (hunk 19 above)
Comment (jackpaparian) [reply bunkerhill/examples/hippocampus/requirements.txt:1]: `nnUNet_predict` requires `matplotlib` to be installed. Though `matplotlib` is listed as a dependency in the current version of `setup.py` on the [master branch](https://github.com/MIC-DKFZ/nnUNet/blob/7f1e273fa1021dd2ff00df2ada781ee3133096ef/setup.py#L26), `matplotlib` is not automatically installed when downloading `nnunet==1.7.0` from PyPI.

The tag for release `1.7.0` is not listed on Github. The latest listed release (`1.6.5`) does not include `matplotlib` as a dependency in [setup.py](https://github.com/MIC-DKFZ/nnUNet/blob/80972d4d79f3337fdfe5fc327821f57e8396779c/setup.py#L11).

File: README.md (hunk 2 above)
Comment (jackpaparian) [reply README.md:128]: I added this section to provide background context for the [Interact with `ModelRunner` server](https://github.com/bunkerhillhealth/bunkerhill-sdk/blob/20e6bce5a2fdd2f429dbd7aa81d6ade211c6c3db/README.md#interact-with-modelrunner-server) section. Happy to restructure this based on user feedback.
